src/wayland/output/mod.rs

```diff
@@ -312,4 +318,34 @@ impl Output {
             .iter()
             .any(|o| o.as_ref().equals(output.as_ref()))
     }
+
+    /// This function allows to run a `Fn` for the matching
+    /// client outputs for a specific `Resource`.
+    pub fn with_output<F, R, II>(&self, resource: &R, mut f: F)
```

Comment:
If the goal is that this closure is invoked for all `wl_output` instances of a given client, why not directly take the `Client` as argument? The user can access the resource in the closure by capturing it. That would make the behavior clearer in my opinion.

Reply:
What do you think about having two functions, one that takes a `Client` and a `FnMut` which will invoke the callback for each output. And a second one to get the first matching output for a resource with the same client.
The second function is more for convenience, like to get a output for a surface when the fullscreen request did not specify an output explicit. While it is possible to use the first function for this use-case i find it really ugly. 

Like:

```
.with_primary(|output, geometry| {
    let primary_output = wl_surface.as_ref().client().and_then(|client| {
        let mut primary_output: Option<wl_output::WlOutput> = None;
        output.with_client_outputs(client, |output| {
            primary_output = Some(output.to_owned());
        });
        primary_output
    });

    (primary_output, geometry)
})
```

vs

```
.with_primary(|output, geometry| {
    let primary_output = output.find(wl_surface);
    (primary_output, geometry)
})
```

Reply:
I think there is a misunderstanding here, because retrieving a single `WlOutput` from smithay's `Output` abstraction is never the correct thing to do.

If there are several `WlOutput` for a single client, that means that client has bound the global sereval times, and in that case all relevant events should be send to all of them. For example if a client has bound a `WlOutput` two times, it should receive the `WlSurface.enter()` event twice, once for each of its `WlOutput` objects that correspond to the same global.

That's what I was interpreting this function to be for: do some processing (like sending an event) to all duplicated relevant output instances. So I don't get what your `primary_output` example is about, a primary output is a notion that exists at the logical output level (and arguably does not really exists in Wayland),but it has very little to do with the `WlOutput` objects...

Reply:
Yeah, definitely a misunderstanding, it is clear to me that an event has to be sent to all bound outputs and I do not argue against it. I see that adding a function to get a single output is wrong in most cases and such things are better prevented at the API level :+1: 

The reason for getting a `WlOutput` is that a client is free to not specify a `WlOutput` when requesting `Fullscreen`. In that
case the compositor should decide if and when yes on which output the window will be rendered fullscreen. As the output map is currently in anvil the only information I found that could be set in the xdg state to remember is a `WlOutput`.
But of course there are other solutions, one thing the comes to my mind is to just leave `None` in the xdg state for the choosen output and remember the output name within anvil and only use the geometry for configure.

For now I will just remove this part of the logic as anvil does not handle fullscreen rendering, so maybe it is better to think about it when implementing it.

Reply:
I renamed the function and changed it to take a client, the provided function is called for each client output.
As written in the previous comment is also removed the strange output behavior. I left the re-configure for fullscreen
by using the output geometry. Everything else for fullscreen handling is a topic for a seperate PR :)